odilo.py

```diff
@@ -0,0 +1,598 @@
+        else:
+            patrons_in_hold_queue = 0
+
+        return CirculationData(
+            data_source=DataSource.ODILO,
+            primary_identifier=primary_identifier,
+            licenses_owned=licenses_owned,
+            licenses_available=licenses_available,
+            licenses_reserved=licenses_reserved,
+            patrons_in_hold_queue=patrons_in_hold_queue,
+        )
+
+    @classmethod
+    def image_link_to_linkdata(cls, link, rel):
+        if not link:
+            return None
+
+        return LinkData(rel=rel, href=link, media_type=Representation.JPEG_MEDIA_TYPE)
+
+    @classmethod
+    def record_info_to_metadata(cls, book, availability):
+        """Turn Odilo's JSON representation of a book into a Metadata
+        object.
+
+        Note:  The json data passed into this method is from a different file/stream
+        from the json data that goes into the book_info_to_circulation() method.
+        """
+        if 'id' not in book:
+            return None
+
+        odilo_id = book['id']
+        primary_identifier = IdentifierData(Identifier.ODILO_ID, odilo_id)
+        active = book.get('active')
+
+        title = book.get('title')
+        subtitle = book.get('subtitle')
+        series = book.get('series')
+        series_position = book.get('seriesPosition')
+
+        contributors = []
+        author = book.get('author')
+        if author:
+            roles = [Contributor.AUTHOR_ROLE]
+            contributor = ContributorData(sort_name=author, display_name=author, roles=roles, biography=None)
+            contributors.append(contributor)
+
+        publisher = book.get('publisher')
+
+        # Metadata --> Marc21 260$c
+        published = book.get('publicationDate')
+        if not published:
+            # yyyyMMdd --> record creation date
+            published = book.get('releaseDate')
+
+        if published:
+            try:
+                published = datetime.datetime.strptime(published, "%Y%m%d")
+            except ValueError as e:
+                cls.log.warn('Cannot parse publication date from: ' + published + ', message: ' + e.message)
+
+        # yyyyMMdd --> record last modification date
+        last_update = book['modificationDate']
+        if last_update:
+            try:
+                last_update = datetime.datetime.strptime(last_update, "%Y%m%d")
+            except ValueError as e:
+                cls.log.warn('Cannot parse last update date from: ' + last_update + ', message: ' + e.message)
+
+        language = book.get('language')
+
+        subjects = []
+        for subject in book.get('subjects', []):
+            subjects.append(SubjectData(type=Subject.TAG, identifier=subject, weight=100))
+
+        grade_level = book.get('gradeLevel')
+        if grade_level:
+            subject = SubjectData(type=Subject.GRADE_LEVEL, identifier=grade_level, weight=10)
+            subjects.append(subject)
```

Comment:
I noticed that the example grade level in your test is 'K-12'. I recognize that 'K-12' is _formatted_ like a grade level, but it covers every single grade in the US school system, making it seem more like an indicator of children-ness than a grade level per se. Is 'K-12' just an unfortunate example? Do you have more specific grade levels like '3-4'?

Reply:
'K-12' is just a metadata example. Yes, we have more specific grade levels as you said.